### PR TITLE
Implement the `has_pending_changes` feature for the views

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

The `has_pending_changes` has been mentioned as a nice-to-have feature for the views. This implements it.

## Proposal

The following was done:
* The `has_pending_changes` has been implemented.
* It has to be `async` since one has to access updates that are protected under a `Mutex`.
* It does not return errors, which is nice.
* Another correction for the `hashable_wrapper` has been done.
* Tests have been added to the containers.

## Test Plan

The tests have been improved to test the feature.

## Release Plan

The feature will need to be added to the documentation.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

**Comment by ma2bd** [comment]
Thanks. I think I still prefer the initially-suggested name "has_pending_changes"

**Comment by MathieuDutSik** [comment]
> Thanks. I think I still prefer the initially-suggested name "has_pending_changes"

Ok, renamed.

File: linera-views/src/hashable_wrapper.rs [hunk 1]
```diff
@@ -40,7 +40,7 @@ impl<C, W, O> View<C> for WrappedHashableContainerView<C, W, O>
 where
     C: Context + Send + Sync,
     ViewError: From<C::Error>,
-    W: HashableView<C> + Send,
+    W: HashableView<C> + Send + Sync,
```
Comment (ma2bd) [inline linera-views/src/hashable_wrapper.rs:43]: Is this intended?

File: linera-views/src/hashable_wrapper.rs [hunk 2]
```diff
@@ -70,25 +70,32 @@ where
...
+
     fn flush(&mut self, batch: &mut Batch) -> Result<bool, ViewError> {
         let delete_view = self.inner.flush(batch)?;
+        let hash = self.hash.get_mut();
         if delete_view {
             let mut key_prefix = self.inner.context().base_key();
             key_prefix.pop();
             batch.delete_key_prefix(key_prefix);
             self.stored_hash = None;
-        } else {
```
Comment (ma2bd) [inline linera-views/src/hashable_wrapper.rs:80]: What is this change about? Can it be in a different PR?

File: linera-views/src/hashable_wrapper.rs (hunk 1 above)
Comment (MathieuDutSik) [reply linera-views/src/hashable_wrapper.rs:43]: Yes.
That is required because of the use of `self.hash.lock().await`.

File: linera-views/src/hashable_wrapper.rs (hunk 2 above)
Comment (MathieuDutSik) [reply linera-views/src/hashable_wrapper.rs:80]: Ok, I have a branch for that. I will create a draft PR once the CI is working again.

File: linera-views/src/key_value_store_view.rs [hunk 3]
```diff
@@ -190,6 +190,26 @@ where
...
+            return true;
+        }
+        if self.sizes.has_pending_changes().await {
+            return true;
+        }
+        let hash = self.hash.lock().await;
+        if self.stored_hash != *hash {
+            return true;
+        }
+        self.stored_total_size != self.total_size
```
Comment (jvff) [inline linera-views/src/key_value_store_view.rs:210]: Optional nit: it might make sense to do all `sync` checks before the `async` checks, just in case they are faster.

File: linera-views/src/views.rs [hunk 4]
```diff
@@ -30,6 +30,9 @@ pub trait View<C>: Sized {
     /// Discards all pending changes. After that `flush` should have no effect to storage.
     fn rollback(&mut self);
 
+    /// Returns true if the flush is going to be non-trivial
```
Comment (jvff) [inline linera-views/src/views.rs:33]: Optional: rephrasing a bit.

```suggestion
    /// Returns [`true`] if flushing this view would result in changes to the persistent storage.
```

File: linera-views/tests/collection_tests.rs [hunk 5]
```diff
@@ -115,8 +116,13 @@ async fn classic_collection_view_check() {
             assert_eq!(key_values, new_map);
         }
         if save {
+            if map != new_map {
+                assert!(view.has_pending_changes().await);
+            }
             map = new_map.clone();
             view.save().await.unwrap();
+            println!("After SAVE");
```
Comment (jvff) [inline linera-views/tests/collection_tests.rs:124]: Question: Is this `println!` intentional?

**Review by jvff — APPROVED** [review]

File: linera-views/src/key_value_store_view.rs (hunk 3 above)
Comment (MathieuDutSik) [reply linera-views/src/key_value_store_view.rs]: Yes, you are right.
